File: app/src/main/java/com/ed/redditapp/ui/postlist/PostListAdapter.java
package com.ed.redditapp.ui.postlist;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.ed.redditapp.R;

import java.util.List;

import lombok.Setter;

public class PostListAdapter extends RecyclerView.Adapter<PostListViewHolder> {
    private static final int VIEW_TYPE_FIRST = 1;
    private static final int VIEW_TYPE_DEFAULT = 0;
    private static final Integer PAYLOAD_ICON_CHANGED = 1;

    public interface ClickListener {
        void onClick(PostListItem post);
    }

    private PostListItem[] posts;

    @Setter private ClickListener infoClickListener;
    @Setter private boolean isDisplayIcon = true;

    @NonNull
    @Override
    public PostListViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.listitem_post, parent, false);
        return new PostListViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull PostListViewHolder holder, int position) {
        PostListItem post = posts[position];
        holder.setTitle(post.getTitle());
        holder.setCommentsCount(post.getCommentsCount());
        holder.setInfo(post.getPoints(), post.getTimestamp(), post.getUsername());
        holder.setThumbnail(post.getLargestThumbnail());
        holder.setSubreddit(post.getSubreddit());
        holder.setSubredditIcon(post.getSubredditIconUrl());
        holder.setDisplayIcon(isDisplayIcon);
        holder.setInfoClickListener(v -> infoClickListener.onClick(post));
    }

    @Override
    public void onBindViewHolder(
            @NonNull PostListViewHolder holder,
            int position,
            @NonNull List<Object> payloads) {
        if (payloads.size() == 0) {
            super.onBindViewHolder(holder, position, payloads);
        } else {
            holder.setSubredditIcon(posts[position].getSubredditIconUrl());
        }
    }

    @Override
    public int getItemCount() {
        if (posts == null) {
            return 0;
        }

        return posts.length;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0)
            return VIEW_TYPE_FIRST;
        else
            return VIEW_TYPE_DEFAULT;
    }

    public void updateData(PostListItem[] posts) {
        this.posts = posts;
        notifyDataSetChanged();
    }

    public void updateIcon(String iconUrl, int position) {
        posts[position].setSubredditIconUrl(iconUrl);
        notifyItemChanged(position, PAYLOAD_ICON_CHANGED);
    }
}

File: app/src/main/java/com/ed/redditapp/ui/postlist/PostListAdapter.kt
package com.ed.redditapp.ui.postlist

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.ed.redditapp.R
import com.ed.redditapp.lib.api.Post

typealias PostListClickListener = (post: Post) -> Unit
typealias PostListNearEndListener = (post: Post) -> Unit

class PostListAdapter: RecyclerView.Adapter<PostListViewHolder>() {
    private val VIEW_TYPE_FIRST = 1
    private val VIEW_TYPE_DEFAULT = 0
    private val PAYLOAD_ICON_CHANGED = 1

    private var posts: Array<Post> = emptyArray()

    var infoClickListener: PostListClickListener? = null
    var titleClickListener: PostListClickListener? = null
    var thumbnailClickListener: PostListClickListener? = null
    var nearEndListener: PostListNearEndListener? = null

    var isDisplayIcon: Boolean = true

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PostListViewHolder {
        val itemView = LayoutInflater
                .from(parent.context)
                .inflate(R.layout.listitem_post, parent, false)
        return PostListViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: PostListViewHolder, position: Int) {
        val post = posts[position]
        holder.setTitle(post.title)
        holder.setCommentsCount(post.commentsCount)
        holder.setInfo(post.points, post.timestamp, post.username)
        holder.setThumbnail(post.largestThumbnail)
        holder.setSubreddit(post.subreddit)
        holder.setSubredditIcon(post.subredditIconUrl)
        holder.setDisplayIcon(isDisplayIcon)
        holder.setInfoClickListener { infoClickListener?.invoke(post) }
        holder.setTitleClickListener { titleClickListener?.invoke(post) }
        holder.setThumbnailClickListener { thumbnailClickListener?.invoke(post) }

        if (position == posts.size - 8) {
            nearEndListener?.invoke(post)
        }
    }

    override fun onBindViewHolder(
            holder: PostListViewHolder,
            position: Int,
            payloads: MutableList<Any>) {
        if (payloads.size == 0) {
            super.onBindViewHolder(holder, position, payloads)
        } else {
            holder.setSubredditIcon(posts[position].subredditIconUrl)
        }
    }

    override fun getItemCount(): Int {
        return posts.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) VIEW_TYPE_FIRST else VIEW_TYPE_DEFAULT
    }

    fun updateData(posts: Array<Post>) {
        this.posts = posts
        notifyDataSetChanged()
    }

    fun updateIcon(iconUrl: String, position: Int) {
        posts[position].subredditIconUrl = iconUrl
        notifyItemChanged(position, PAYLOAD_ICON_CHANGED)
    }

    fun addPosts(newPosts: Array<Post>) {
        val positionBefore = posts.size - 1
        posts += newPosts
        notifyItemRangeInserted(positionBefore, newPosts.size)
    }
}